### PR TITLE
soc: riscv: cleanup machine timer related code

### PR DESCRIPTION
**Pull request by gmarull**
After some analysis I found out that there's no machine timer provided
by the "riscv" vendor, so the existing "riscv,machine-timer" binding was not
designed to fit all vendors. There are some specs for the mtime/mtimecmp
registers (this is why we can have a single driver), but the actual
register layout or implementations differ among vendors. One of them is CLINT
(Core Local Interruptor) which includes machine timer registers. A few platforms already
defined it correctly, but it was never used to obtain mtime/mtimecmp base address nor interrupts.

This patch moves all the information related to machine timer in Devicetree (aligning with Linux in some cases) and removes definitions from soc.h.

Supersedes https://github.com/zephyrproject-rtos/zephyr/pull/48382
Part of RISC-V soc.h cleanup/removal.

### CODE REVIEW

**Review by carlocaione — DISMISSED** [review]
This in general LGTM.

The only problem I see is that probably CLINT would have required a different driver class (maybe syscon instead of timer?) because it really is a bag of mixed nuts. You have registers for timer, registers for user-generated interrupts and even registers for IPI in the same MMIO region.

**Comment by gmarull** [comment]
> This in general LGTM.
> 
> The only problem I see is that probably CLINT would have required a different driver class (maybe syscon instead of timer?) because it really is a bag of mixed nuts. You have registers for timer, registers for user-generated interrupts and even registers for IPI in the same MMIO region.

I agree that this design isn't great, the fact that each vendor places mtime/mtimecmp in different places doesn't help to find a nice solution. The lack of proper DT description really hid the problem or the complexity.

**Review by carlocaione — DISMISSED** [review]

**Review by nandojve — DISMISSED** [review]
Just a remind that there is the CLIC extension which is compatible with CLINT to keep in radar. It is already available in some SoC like SiFive E24. Specs for reference at https://github.com/riscv/riscv-fast-interrupt

File: dts/bindings/timer/litex,machine-timer.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  LiteX Machine Timer
+
+  The LiteX machine timer provides RISC-V prigileved mtime and mtimecmp
+  registers.
```
Comment (fkokosinski) [inline dts/bindings/timer/litex,machine-timer.yaml:8]: `litex_vexriscv` doesn't fully follow the RISC-V privileged spec and doesn't use the RISC-V machine timer driver.

File: dts/bindings/timer/andestech,machine-timer.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Andes Machine Timer
+
+  The Andes machine timer provides RISC-V prigileved mtime and mtimecmp
```
Comment (fkokosinski) [inline dts/bindings/timer/andestech,machine-timer.yaml:7]: prigileved -> privileged (here and in the rest of bindings)

File: dts/riscv/riscv32-litex-vexriscv.dtsi [hunk 3]
```diff
@@ -42,6 +42,14 @@
 			reg-names = "irq_mask", "irq_pending";
 			riscv,max-priority = <7>;
 		};
+
+		mtimer: timer@ffffffe0 {
+			compatible = "litex,machine-timer";
+			reg = <0xffffffe0 0x10>;
+			interrupts = <7 0>;
+			interrupt-parent = <&intc0>;
+		};
```
Comment (fkokosinski) [inline dts/riscv/riscv32-litex-vexriscv.dtsi:51]: Ditto - this node shouldn't be added for `litex_vexriscv`.

**Review by fkokosinski — CHANGES_REQUESTED** [review]

File: dts/bindings/timer/litex,machine-timer.yaml (hunk 1 above)
Comment (fkokosinski) [reply dts/bindings/timer/litex,machine-timer.yaml:8]: Hmm, I don't remember why I had added that line of code 3 years ago when implementing the initial irqc driver, but `litex_vexriscv` doesn't use this IRQ at all (tested on HW). So I'd suggest just removing the reference to `RISCV_MACHINE_TIMER_IRQ` and just leaving the `MACHINE_EXT` one.

File: dts/bindings/timer/andestech,machine-timer.yaml (hunk 2 above)
Comment (gmarull) [reply dts/bindings/timer/andestech,machine-timer.yaml]: fixed on all bindings, thanks

File: dts/bindings/timer/litex,machine-timer.yaml (hunk 1 above)
Comment (gmarull) [reply dts/bindings/timer/litex,machine-timer.yaml]: removed

**Review by carlocaione — APPROVED** [review]

**Review by cameled — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

File: drivers/timer/riscv_machine_timer.c [hunk 4]
```diff
@@ -4,10 +4,47 @@
...
  */
 #include <zephyr/device.h>
+#include <zephyr/devicetree.h>
 #include <zephyr/drivers/timer/system_timer.h>
 #include <zephyr/sys_clock.h>
 #include <zephyr/spinlock.h>
-#include <soc.h>
+
+/* neorv32-machine-timer */
+#if DT_HAS_COMPAT_STATUS_OKAY(andestech_machine_timer)
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/timer/riscv_machine_timer.c:13]: oops will now send a follow-up PR

**Pull request merged**